internal/handlers/common/update.go

```diff
@@ -746,6 +752,91 @@ func processCurrentDateFieldExpression(doc *types.Document, currentDateVal any)
 	return changed, nil
 }
 
+func processBitFieldExpression(command string, doc *types.Document, updateV any) (bool, error) {
+	var changed bool
+
+	bitDoc := updateV.(*types.Document)
+
+	for _, bitKey := range bitDoc.Keys() {
+		bitValue := must.NotFail(bitDoc.Get(bitKey))
+
+		nestedDoc, ok := bitValue.(*types.Document)
+		if !ok {
+			return false, newUpdateError(
+				commonerrors.ErrTypeMismatch,
+				fmt.Sprintf(
+					`The $bit modifier is incompatible with a %s. `+
+						`You must pass in an embedded document: {$bit: {field: {and/or/xor: #} }}`,
+					commonparams.AliasFromType(bitValue),
+				),
+				command,
+			)
+		}
+
+		path := must.NotFail(types.NewPathFromString(bitKey))
+
+		// $bit ignores fields that does not exist
+		if !doc.HasByPath(path) {
+			continue
+		}
+
+		docValue, err := doc.GetByPath(path)
+		if err != nil {
+			return false, err
+		}
+
+		// $bit operations can only be performed on integer fields
+		switch docValue.(type) {
+		case int32, int64:
+		default:
+			return false, newUpdateError(
+				commonerrors.ErrTypeMismatch,
+				fmt.Sprintf(
+					`Cannot apply $bit to a value of non-integral type. `+
+						`{_id: %s} has the field '%s' of non-numeric type %s`,
+					types.FormatAnyValue(must.NotFail(doc.Get("_id"))),
+					bitKey,
+					commonparams.AliasFromType(docValue),
+				),
+				command,
+			)
+		}
+
+		for _, bitOp := range nestedDoc.Keys() {
+			bitOpValue := must.NotFail(nestedDoc.Get(bitOp))
+			fieldType := commonparams.AliasFromType(bitOpValue)
+
+			// $bit operator argument must be an integer (Int32/Long)
+			if fieldType != "int" && fieldType != "long" {
+				return false, newUpdateError(
+					commonerrors.ErrTypeMismatch,
+					fmt.Sprintf(
+						`The $bit modifier field must be an Integer(32/64 bit); a `+
+							`%s is not supported here`,
+						fieldType,
+					),
+					command,
+				)
+			}
+
+			bitOpResult, err := performBitLogic(bitOp, bitOpValue, docValue)
+			if err == nil {
+				if err = doc.SetByPath(path, bitOpResult); err != nil {
+					return false, lazyerrors.Error(err)
+				}
+
+				changed = true
+
+				continue
+			}
+
+			return false, err
```

Comment:
We want to return error with appropriate message here. For example, what if the operator was not `and`, `or` or `xor`?

Reply:
Is `commonerrors.ErrInvalidArg` an appropriate error code for this use case. I'm not sure of the error code to use

Reply:
In general, we should return the same error code as MongoDB. What does it produce in that case? It should be easy to check with FerretDB's proxy mode.

Reply:
Indeed we want to return the same error code as MongoDB :hugs:. Also our compatibility tests check error codes are the same, so we should add test case for failing case and ensure that test passes.